### PR TITLE
Provide unit tests for NotificationManager, NotitificationEmail and default message/subject

### PR DESCRIPTION
**Pull request by mchandschuh**
#### Description

Without access to the `IMessagingHandler` implementation used in the production environment, the next best solution is to provide a default value for `message` and `subject` to ensure it's never null. The likely culprit is a null reference exception that is being swallowed on a different thread where the notification messages are being processed.

This change adds unit tests for `NotificationManager` as well as provides some basic refactoring to match style guidelines, stream-line `Allow` checks, as well as removing a race condition in the rate limiting checks (perhaps should be updated to `RateGate`, but felt it was outside the scope of this PR). Additionally, the `_resetTime` was changed to use `DateTime.UtcNow` to avoid potential issues with local time zone settings and the inevitable time discontinuity that arises from things like daylight savings adjustments.

In addition to providing default values for `NotificationEmail.Subject` and `NotificationEmail.Message`, validation has been added to ensure a valid email address has been provided via a new `Validate` class with a `Validate.EmailAddress` method. The reasoning behind a new class here is to provide a location for further string based validations (such as phone number, web address, etc) and to provide a singular location for potentially complicated regular expressions.

Another change was made to the `QuantConnect.sln.DotSettings` file to include `QUANTCONNECT` in the custom dictionary to avoid R# typo warnings in the license headers.

#### Related Issue

See #3021 - Notify.Email fails silently if message is null

#### Motivation and Context

The `Notify.Email` method should not fail when given a null message or subject as these are mandatory fields when sending an email.

#### Requires Documentation Change

No. This change updates the behavior to the expected behavior of not failing.

#### How Has This Been Tested?

Changes extensively tested via unit tests.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Tests/Common/Notifications/NotificationEmailTests.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+﻿using System;
```
Comment (Martin-Molinero) [inline Tests/Common/Notifications/NotificationEmailTests.cs]: New file missing header

File: Common/Notifications/Notification.cs [hunk 2]
```diff
@@ -119,10 +124,15 @@ public class NotificationEmail : Notification
...
-            Subject = subject;
+            if (!Validate.EmailAddress(address))
+            {
+                throw new ArgumentException($"Invalid email address: {address}");
+            }
+
             Address = address;
+            Data = data ?? string.Empty;
+            Message = message ?? DefaultMessageAndSubject;
+            Subject = subject ?? DefaultMessageAndSubject;
```
Comment (Martin-Molinero) [inline Common/Notifications/Notification.cs]: Maybe it's worth documenting this new behavior in the methods xml docs? If null will use `DefaultMessageAndSubject/string.Empty`

File: Common/Util/Validate.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                return null;
+            }
+
+            return emailAddress;
+        }
+
+        /// <summary>
+        /// Provides static storage of compiled regular expressions to preclude parsing on each invocation
+        /// </summary>
+        public static class RegularExpression
```
Comment (Martin-Molinero) [inline Common/Util/Validate.cs]: nit: wondering if it's worth for this guy to be public and not private, thoughts?

Comment (mchandschuh) [reply Common/Util/Validate.cs]: I had gone back and forth on this one and figured exposing it doesn't do any harm. The added benefit to exposing it is there can be capture groups in the regular expressions that can be useful depending on what you're doing. We would need to expose methods for fetching the capture groups for each regex. This behavior isn't currently used, but was the reasoning. I'm happy to move it to private as well.

File: Common/Notifications/Notification.cs (hunk 2 above)
Comment (mchandschuh) [reply Common/Notifications/Notification.cs]: Agreed. Notify docs should be updated as well as this constructor's documentation.

File: Tests/Common/Notifications/NotificationEmailTests.cs (hunk 1 above)
Comment (mchandschuh) [reply Tests/Common/Notifications/NotificationEmailTests.cs]: will fix.

**Comment by mchandschuh** [comment]
> 
> `Notify.Email` fails with a system log at https://github.com/QuantConnect/Lean/blob/master/Engine/Results/LiveTradingResultHandler.cs#L1120 . Wondering if these failures should reach the users through a log/message in the cloud console, maybe through https://github.com/QuantConnect/Lean/blob/master/Engine/Results/LiveTradingResultHandler.cs#L450 ? Thinking about the other type of notification failures and other potential failures with the email type.

Great point. We should likely be exposing these to the user via the algorithm's logging facilities (or via result handler). Are there any concerns regarding leakage of details of code outside of this repository. For example, in production, we run a custom messaging handler in the cloud, so maybe that implementation needs to be updated to catch errors and raise them in a user-friendly way (perhaps there's a dedicated error type that could hold a user-friendly message)

File: Common/Util/Validate.cs (hunk 3 above)
Comment (Martin-Molinero) [reply Common/Util/Validate.cs]: Agree, there is no one right choice here, personally, I'd keep it private until required but it's good as it too 👍

**Comment by Martin-Molinero** [comment]
> Great point. We should likely be exposing these to the user via the algorithm's logging facilities (or via result handler). Are there any concerns regarding leakage of details of code outside of this repository. For example, in production, we run a custom messaging handler in the cloud, so maybe that implementation needs to be updated to catch errors and raise them in a user-friendly way (perhaps there's a dedicated error type that could hold a user-friendly message)

Good point! Yeah, agree, it's more of a cloud error handling responsability 👍, so not for this PR. The cloud messaging handler is actually handling errors for custom notifications, should just be expanding it to cover all type of notifications.

**Comment by jaredbroad** [comment]
Lots of nice validation goodies in this PR but would appreciate if we could pull them to QCAlgorithm.Notify.Email() as input validation steps instead of trying to solve deep in LEAN? A throw on a bad email address would easily reach user if inside API layer/Notify.Email() method. 

Recommend forcing good inputs instead of supporting nulls. Force the 1% of people who want null subjects to use string.Empty / set their own defaults.

**Comment by mchandschuh** [comment]
@jaredbroad RE: validation in Notify.Email - - isn't that where it is? This PR adds validation to the notify email constructor which is invoked via the Notification Manager which in turn is directly exposed via IAlgorithm [here](https://github.com/QuantConnect/Lean/blob/0093501af4a2defff5fed0173cbc73187419991b/Common/Interfaces/IAlgorithm.cs#L128)

**Pull request merged**